smk/include/bwa_index_wrapper.smk

```diff
@@ -0,0 +1,226 @@
+#    input:  {somewhere}/BWA_index/{something}.fna.*
+#    output: CLUSTER_LOCAL_DIR/{somewhere}/BWA_index/{something}.fna.*
+# 3. Symlinking to local drive:
+#    input:  CLUSTER_LOCAL_DIR/{somewhere}/BWA_index/{something}.fna.*
+#    output: {somewhere}/BWA_index_local/{something}.fna.*
+#
+# Usage:
+# ------
+# When using this wrapper, a rule should figure out whether it will be executed locally or on a cluster.
+# Then it should ask for bwaindex files as follows:
+#   {somewhere}/BWA_index/{something}.fna.*:       if the files can be read directly without staging.
+#   {somewhere}/BWA_index_local/{something}.fna.*: if the files need to be staged to all nodes in cluster.
+#
+# See rule 'genome_mapping_profiling' in gene_abundance.smk for usage example.
+#
+# Note on CLUSTER_NODES:
+# ----------------------
+# Values in the python array CLUSTER_NODES will be used as node names to distribute the files.
+# This is easier to control when run with '--cluster' since the rule will submit the job to the
+# CLUSTER_WORKLOAD_MANAGER with a request to run only on that node.
+# If it is not run with '--cluster', by mistake, this could cause trouble: now all jobs are run
+# on the same node where you invoked your snakemake command. And this node could wrongly create
+# a 'status' file that denotes sync has been done on a different node. We have a simple solution for this.
+# The node names in CLUSTER_NODES should match the node name accessible by the command 'hostname --short'.
+# The sync status files are named with the local node name. You ask for sync'ing in nodeX, and
+# only nodeX can make that file. This avoids problems when this is NOT run using --cluster.
+# There is anothe problem that multiple sync rules may run the same 'rsync source dest' command
+# on a single server. This would likely cause a race condition. To avoid that, we create a lock
+# using 'flock' so that only one rsync is running for a specific bwa index sync on one node.
+# As a bonus, this also allows rsyncing an NFS bwa index onto a local disk when running without
+# --cluster, if you define CLUSTER_NODES with just your exec node's name.
+################################################################################################
+
+if CLUSTER_WORKLOAD_MANAGER is not None:
+    known_managers = ['SLURM', 'PBS', 'TORQUE', 'SGE', 'LSF']
+    if CLUSTER_WORKLOAD_MANAGER.upper() not in known_managers:
+        raise Exception(f"MIntO error: Unexpected value: CLUSTER_WORKLOAD_MANAGER='{CLUSTER_WORKLOAD_MANAGER}'. Must be one of {known_managers}")
+
+if CLUSTER_NODES is not None:
+
+    # Make sure other variables are defined
+    if CLUSTER_WORKLOAD_MANAGER is None or CLUSTER_LOCAL_DIR is None:
+        raise Exception(f"MIntO error: CLUSTER_WORKLOAD_MANAGER and CLUSTER_LOCAL_DIR must be defined, if CLUSTER_NODES is defined")
+
+    # Baseline    : 5 GB
+    # Size-based  : 22 byte per byte file size
+    # New attempts: +40 GB each time
+    rule BWA_index_contigs_nfs:
```

Comment:
Isn't this meant to be a shadow rule?

Reply:
Not this rule. This is to save traffic. The idea is that one node makes the index, pushes to NFS and then all nodes rsync it to their local disk. But the maker node already has it in its local disk if we used shadow rule, so why rsync a second time? So instead of shadow-dir, we make it directly at the final location on local disk. Workflow will run `rsync` on all `N` nodes, but the maker node will check that files are already up-to-date and will not transfer the data. So `rsync` only transfers data on `N-1` nodes.